roles/smartgateway/templates/metrics-configmap.yaml.j2

```diff
@@ -4,9 +4,9 @@ metadata:
   name: {{ meta.name }}-smartgateway
   namespace: {{ meta.namespace }}
 data:
-  "sa.metrics.config.json": |
+  "smartgateway_config.json": |
     {
-            "AMQP1MetricURL": "{{ amqpmetric_url | default('messaging-internal-{{ meta.name }}.{{ meta.namespace }}.svc:5672/collectd/telemetry') }}",
+            "AMQP1MetricURL": "{{ amqp_url | default('messaging-internal-' + meta.name + '.' + meta.namespace + '.svc:5672/collectd/telemetry') }}",
```

Comment:
In the Helm world we try to avoid the overuse of defaults in the templates, and instead try to explicitly declare every template variable with a default (in roles/smartgateway/defaults/main.yml in this case). This leads to fewer conditionals in the templates ("if defined" kind of stuff) as well as separating default config from the template logic itself. The final major benefit is that it more explicitly documents the "API" of your playbook - without setting a default in main.yml, the only way I'd even know that I can supply an amqp_url variable would be by reading every template carefully.

Does this suggestion apply here, or is there some reason that declaring your defaults and API in one place isn't a good choice in Ansible?

Reply:
So the reason I have the defaults here, is I don't think (could be wrong, will investigate) that I can define the variables in the `defaults/main.yml`, i.e. I can't make the value of the default dynamic. I will double check, because if I can do that, then I definitely would prefer to remove the defaults out of the template here (which makes it easier to read) and have a full default set-list in a single file. Definitely makes it more self-documenting.